tests/unit/lms/services/digest_test.py

```diff
@@ -0,0 +1,510 @@
+
+from lms.services.digest import (
+    DigestHelper,
+    DigestService,
+    UnifiedCourse,
+    UnifiedUser,
+    UnknownAuthorityProvidedID,
+    service_factory,
+)
+from lms.services.mailchimp import EmailRecipient, EmailSender
+from tests import factories
+
+
+class Annotation(factory.Factory):
+    """
+    A factory for annotation dicts.
+
+    >>> Annotation()
+    {'author': {'userid': 'acct:user_1@lms.hypothes.is'}, 'group': {'authority_provided_id': 'group_1'}}
+    >>> Annotation(userid='acct:custom_username@lms.hypothes.is')
+    {'author': {'userid': 'acct:custom_username@lms.hypothesis.is'}, 'group': {'authority_provided_id': 'group_2'}}
+    >>> Annotation(authority_provided_id='custom_id')
+    {'author': {'userid': 'acct:user_2@lms.hypothes.is'}, 'group': {'authority_provided_id': 'custom_id'}}
+    """
+
+    class Meta:
+        model = dict
+
+    userid = factory.Sequence(lambda n: f"acct:user_{n}@lms.hypothes.is")
+    authority_provided_id = factory.Sequence(lambda n: f"group_{n}")
+
+    @factory.post_generation
+    def post(obj, *_args, **_kwargs):  # pylint:disable=no-self-argument
+        # pylint:disable=unsupported-assignment-operation,no-member
+        obj["author"] = {"userid": obj.pop("userid")}
+        obj["group"] = {"authority_provided_id": obj.pop("authority_provided_id")}
+        return obj
+
+
+class TestDigestService:
+    def test_send_instructor_email_digests(
+        self, svc, h_api, helper, UnifiedUser_, db_session, mailchimp_service, sender
+    ):
+        h_userids = [sentinel.userid1, sentinel.userid2]
+        unified_users = UnifiedUser_.make.side_effect = [
+            create_autospec(UnifiedUser, spec_set=True, instance=True),
+            create_autospec(UnifiedUser, spec_set=True, instance=True),
+        ]
+        digests = helper.instructor_digest.side_effect = [
+            {"num_annotations": 1},
+            {"num_annotations": 2},
+        ]
+
+        svc.send_instructor_email_digests(h_userids, sentinel.since, sentinel.until)
+
+        h_api.get_annotations.assert_called_once_with(
+            h_userids, sentinel.since, sentinel.until
+        )
+        helper.unified_courses.assert_called_once_with(
+            h_api.get_annotations.return_value
+        )
+        assert UnifiedUser_.make.call_args_list == [
+            call(db_session, h_userid) for h_userid in h_userids
+        ]
+        assert helper.instructor_digest.call_args_list == [
+            call(unified_user, helper.unified_courses.return_value)
+            for unified_user in unified_users
+        ]
+        assert mailchimp_service.send_template.call_args_list == [
+            call(
+                "instructor-email-digest",
+                sender,
+                recipient=EmailRecipient(unified_user.email, unified_user.display_name),
+                template_vars=digest,
+            )
+            for unified_user, digest in zip(unified_users, digests)
+        ]
```

Comment:
This is a weird test but all the actual logic has been tested in the unit tests for the different components below. `send_instructor_email_digests()` is mostly just glue code.